### ISSUE TITLE
no variability in predictions for rlasso on toy example

### ISSUE THREAD

**Issue by bertillepicard** [xnie/rlearner#4]
I use the rlearner package to estimate the CATE function in a RCT. By setting the seed to 1 and testing the toy example, I realize that there is no variability in the predictions. It seems that no variable is therefore selected by the lasso. I also have this problem on real data for which an impact should be detected. Is it related to the length of the glmnet hyperparameter tuning grid? Thanks.

```
library(rlearner) 
set.seed(1)

n = 100; p = 10

x = matrix(rnorm(n*p), n, p)
w = rbinom(n, 1, 0.5)
y = pmax(x[,1], 0) * w + x[,2] + pmin(x[,3], 0) + rnorm(n)

rlasso_fit = rlasso(x, w, y)
rlasso_est = predict(rlasso_fit, x, k_folds=3)
rlasso_est
```

```

         [,1]
  [1,] 0.7918391
  [2,] 0.7918391
  [3,] 0.7918391
  [4,] 0.7918391
  [5,] 0.7918391
  [6,] 0.7918391
  [7,] 0.7918391

```

**Comment by eliozhang-research** [comment]
The issue will be resolved if you change the coefficient of x[,2] to 10.
Yongli

**Comment by eliozhang-research** [comment]
Another "solution" is to use rnorm(n)/2.

Yongli

**Comment by bertillepicard** [comment]
My question was not clear enough, sorry.

My problem is not finding a DGP that works. Rather, the question is, why can't the rlearner give variability for the treatment effect? On this DGP, the xlearner for example manages to give variability. Does this toy example show the limits of the rlearner on small samples, or on a CATE function that is too non-linear? or too simple?

**Comment by xnie** [comment]
The rlearner doesn't necessarily always give variability for the treatment effect, and this can happen for many reasons, e.g. the sample size is too small and the regularization in glmnet (in the case of using rlasso) makes the treatment effect estimate to be constant. The example code provided is merely a a toy DGP example to show how to call the functions. If you are using the lasso version of rlearner, you might want to do a basis (e.g. spline) expansion to the covariates to encode nonlinearities. If you decide to use other machine learning models to plug into the rlearner, I just wanted to point out that the grf package on cran runs the R-learner under the hood with random forests instead of lasso or boosting (as shown in this repo), and I believe they have a function that returns variable importance as well.

**Comment by bertillepicard** [comment]
Thank you for this answer,

I can understand that this lack of variability on the toy example with the lasso can indeed be adjusted by adding expansions.

I tried to change the DGP a little to see how the rlasso reacted. And I also implemented the rboost. For the moment, I'm surprised by the results for the rboost and the difference between the quality of the rlasso and the rboost on this example. Is it due to the scale of this new example?

I may not have truly understood the usage of rboost? 

```
set.seed(2)

library(rlearner)
n = 1000; p = 10

x = matrix(rnorm(n*p,2000,50), n, p)
w = rbinom(n, 1, 0.5)
y = pmax(x[,1], 0) * w + x[,2] + pmin(x[,3], 0) + rnorm(n)

rlasso_fit = rlasso(x, w, y)
rlasso_est = predict(rlasso_fit, x)
rlasso_est

rboost_fit = rboost(x, w, y)
rboost_est = predict(rboost_fit, x)
rboost_est

rlasso_est<- sort(rlasso_est)
rboost_est<-sort(rboost_est)


X1 <- sort(pmax(x[,1], 0))

plot(1:n, X1, type = "b", frame = FALSE, pch = 19, 
     col = "grey", xlab = "x", ylab = "y",ylim=c(0,2100))


lines(1:n, rboost_est, pch = 18, col = "blue", type = "b", lty = 2)

lines(1:n, rlasso_est, pch = 18, col = "yellow", type = "b", lty = 2)

```

In grey, I plotted the true conditional treatment effect which is most of the time in this DGP always positive, so the lasso (yellow) works well. However, rboost (blue) is really far.

![image](https://user-images.githubusercontent.com/52250335/97728392-cc1ef500-1ad1-11eb-85b5-66d202e1c049.png)

**Comment by xnie** [comment]
Thanks a lot for the follow-up, and interesting!

I played around a bit with your DGP setup, and it seems like boosting doesn't seem to do a good job on learning \hat{m} = E[Y|X]. You can see the learned value of \hat{m} by calling `rboost_est$m_hat`. I tried using the lasso learned \hat{m} and hat{e} = E[W|X] instead for r-boosting, by calling `rboost(x, w, y, p_hat = rlasso_fit$p_hat, m_hat = rlasso_fit$m_hat)`, the result appears slightly better. 

In general, we want to learn as accurate propensity score \hat{e} and outcome function \hat{m} as we can (using any machine learning method) for the R-learner. The XGBoost package the rboost function depends on has many tuning parameters that one can adjust, and how well we can minimize the loss function on \hat{e}, \hat{m} and the R-loss depends on whether the machine learning method is well suited (and well tuned) for the problem.

**Comment by bertillepicard** [comment]
Thanks a lot for your answer!

Indeed it is better with the lasso predictions for e and m. The fact that the approximation of boosting is that far from the true y is a bit strange though, even if the model is not well suited. For example, I fitted E[Y|X=x] with xgboost, cross-validating on 5 folds using caret, with the default grid, and I get the distribution in pink. I did not notice a particular increase in the computation time, so I do not think that I gave much more chances to the second boosting method - but I did not precisely measured it. I also plotted the lasso distribution and the true distribution of y. The true y is in two parts, since its value depends on whether the individual is treated or not. Souldn't we get something quite similar for the pink (boost) and green (mhat from rboost fit) distributions?

Then I thought, maybe the default tuning grid was too specific to fit this Y, so I just replaced the "param" object in the cvboost script by 
```
        param <- list(objective = objective, eval_metric = eval)
```

that is, removing your default grid of parameters to let xgboost choosing it. The resulting m_hat looks better I think.

```
library(tidyverse)
library(xgboost)
library(caret)

xgb_trcontrol = trainControl(method = "cv", number = 5, allowParallel = TRUE, 
                             verboseIter = FALSE, returnData = TRUE, 
                             savePredictions = "final")

xgb_model = train(data.frame(x), y, trControl = xgb_trcontrol, tuneGrid = NULL, 
                  method = "xgbTree")

xgb_fit <- xgb_model$pred$pred


y.dataframe <- data.frame(y)
colnames(y.dataframe) <- "m_hat"
y.dataframe$model <- "true y"

lasso.dataframe <- data.frame(rlasso_fit$m_hat)
colnames(lasso.dataframe) <- "m_hat"
lasso.dataframe$model <- "Rlasso"

boost.dataframe <- data.frame(rboost_fit$m_hat)
colnames(boost.dataframe) <- "m_hat"
boost.dataframe$model <- "Rboost"

boost.normal.dataframe <- data.frame(xgb_fit)
colnames(boost.normal.dataframe) <- "m_hat"
boost.normal.dataframe$model <- "boost"

values <- rbind(y.dataframe, lasso.dataframe, boost.dataframe, boost.normal.dataframe)

ggplot(values, aes(x =m_hat, fill = model)) + 
  xlim(min(values$m_hat)-100, max(values$m_hat)+100) + 
  geom_histogram(alpha = 0.5, position = 'identity')
```

![image](https://user-images.githubusercontent.com/52250335/97995472-d64d3600-1de6-11eb-9ab9-b1756d7f72c6.png)

After removing the grid in the cvboost script:

![image](https://user-images.githubusercontent.com/52250335/98016883-cb9e9b00-1dfe-11eb-8c87-ce0ec335c9ba.png)

Then I plotted the same graph as last time:

![image](https://user-images.githubusercontent.com/52250335/98018441-d65a2f80-1e00-11eb-9929-99755ca8760c.png)

**Comment by xnie** [comment]
Thanks for sharing your results and insights! This is very helpful. Indeed hyperparameter choices matter quite a bit for a method such as boosting. The default parameters we have left in the code for the boosting based learners are rather generic, and should be looked at more closely for specific application needs. Having a good prediction power for \hat{e} and \hat{m} is crucial for the R-learner, and that boils down to a standard prediction accuracy tuning problem in machine learning. 

I will close this issue for now, but please feel free to get in touch again (on Github or via my email). Thanks again for bringing this to our attention!
